File: agent/tools.py
"""
Pydantic AI tool definitions for the WODpilot coach agent.

All tools receive the user_id via RunContext dependencies.
The agent autonomously decides which tools to call and in what order.
"""
from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from pydantic_ai import RunContext  # type: ignore

from services.garmin import get_training_load, get_recent_activities
from services.scraper import get_todays_wods
from memory.semantic import search_memory, save_memory, get_all_memories
from memory.episodic import search_episodes, add_episode, get_prs
from memory.procedural import get_coaching_profile, update_coaching_style

logger = logging.getLogger(__name__)


@dataclass
class CoachDeps:
    user_id: int
    user_name: str


# ─── Tool implementations ─────────────────────────────────────────────────────

async def tool_get_training_load(ctx: RunContext[CoachDeps]) -> str:
    """
    Get current ATL/CTL/TSB training load metrics and a recommendation.
    Always call this first when the athlete asks about training readiness.
    """
    load = get_training_load(ctx.deps.user_id)
    return json.dumps(
        {
            "atl": load.atl,
            "ctl": load.ctl,
            "tsb": load.tsb,
            "weekly_tss": load.weekly_tss,
            "run_km_14d": load.run_km_14d,
            "recommendation": load.recommendation,
        },
        ensure_ascii=False,
    )


async def tool_get_recent_activities(ctx: RunContext[CoachDeps], days: int = 7) -> str:
    """
    Get recent Garmin activities for the last N days.
    Use this to understand what the athlete has been doing recently.
    Returns activity type, duration, HR, distance, TSS.
    Warns if running volume > 50 km/14d.
    """
    activities = get_recent_activities(ctx.deps.user_id, days=min(days, 30))
    return json.dumps(activities, ensure_ascii=False, default=str)


async def tool_get_todays_wods(ctx: RunContext[CoachDeps]) -> str:
    """
    Get today's WODs from all configured CrossFit boxes.
    Use this when generating a training recommendation for today.
    """
    wods = get_todays_wods()
    if not wods:
        return "Keine WODs für heute verfügbar. Generiere ein individuelles Workout."
    return json.dumps(wods, ensure_ascii=False, default=str)


async def tool_search_memory(ctx: RunContext[CoachDeps], query: str) -> str:
    """
    Search the athlete's semantic memory (PRs, injuries, goals, preferences).
    Use this to recall relevant facts about the athlete.
    Example queries: 'shoulder injury', 'back squat PR', 'goals 2025'
    """
    results = search_memory(ctx.deps.user_id, query)
    if not results:
        return f"Keine Einträge zu '{query}' gefunden."
    return json.dumps(results, ensure_ascii=False)


async def tool_search_episodes(ctx: RunContext[CoachDeps], query: str) -> str:
    """
    Search episodic memory for specific events (PRs, achievements, injuries).
    Use this when the athlete references past performances.
    """
    results = search_episodes(ctx.deps.user_id, query)
    if not results:
        return f"Keine Episoden zu '{query}' gefunden."
    return json.dumps(results, ensure_ascii=False, default=str)


async def tool_get_all_prs(ctx: RunContext[CoachDeps]) -> str:
    """
    Get all Personal Records for the athlete.
    Use this when discussing performance history or setting new training goals.
    """
    prs = get_prs(ctx.deps.user_id)
    if not prs:
        return "Noch keine PRs gespeichert."
    return json.dumps(prs, ensure_ascii=False, default=str)


async def tool_save_memory(
    ctx: RunContext[CoachDeps],
    key: str,
    value: str,
    category: Optional[str] = None,
) -> str:
    """
    Save an important fact to semantic memory.
    Use this when learning something important about the athlete:
    injuries, PRs, goals, preferences, scaling levels.
    Categories: 'injury', 'pr', 'goal', 'preference', 'scaling'
    """
    save_memory(ctx.deps.user_id, key, value, category)
    return f"Gespeichert: {key} = {value}"


async def tool_add_episode(
    ctx: RunContext[CoachDeps],
    content: str,
    category: Optional[str] = None,
    metadata: Optional[Dict[str, Any]] = None,
) -> str:
    """
    Add an episodic memory entry for a specific event.
    Use this for: PRs, injuries, achievements, notable workouts.
    Categories: 'pr', 'injury', 'achievement', 'scaling', 'wod_result'
    Metadata example: {"exercise": "deadlift", "weight_kg": 145, "reps": 1}
    """
    add_episode(ctx.deps.user_id, content, category, metadata)
    return f"Episode gespeichert: {content}"


async def tool_get_coaching_profile(ctx: RunContext[CoachDeps]) -> str:
    """
    Get the athlete's coaching style profile.
    Use this to adapt communication style (direct/supportive/technical/balanced).
    """
    profile = get_coaching_profile(ctx.deps.user_id)
    return json.dumps(profile, ensure_ascii=False)


async def tool_update_coaching_style(
    ctx: RunContext[CoachDeps],
    coaching_style: Optional[str] = None,
    notes: Optional[str] = None,
) -> str:
    """
    Update the coaching style based on athlete feedback.
    Valid styles: 'direct', 'supportive', 'technical', 'balanced'
    Use 'notes' to remember specific communication preferences.
    """
    update_coaching_style(
        ctx.deps.user_id,
        coaching_style=coaching_style,
        notes=notes,
    )
    return f"Coaching-Stil aktualisiert: {coaching_style or 'unverändert'}"

File: pi-agent-service/src/tools.ts
import { Type } from "typebox";
import type { AgentTool, AgentToolResult } from "@mariozechner/pi-agent-core";

const PYTHON_API = process.env.PYTHON_API_URL ?? "http://web:8080";
const INTERNAL_TOKEN = process.env.INTERNAL_API_TOKEN ?? "";

function makeHeaders(): Record<string, string> {
  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${INTERNAL_TOKEN}`,
  };
}

function textResult(text: string): AgentToolResult<Record<string, never>> {
  return { content: [{ type: "text", text }], details: {} };
}

async function apiGet(path: string): Promise<string> {
  const res = await fetch(`${PYTHON_API}${path}`, { headers: makeHeaders() });
  if (!res.ok) throw new Error(`Tools API ${path} returned ${res.status}`);
  return res.text();
}

async function apiPost(path: string, body: unknown): Promise<string> {
  const res = await fetch(`${PYTHON_API}${path}`, {
    method: "POST",
    headers: makeHeaders(),
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Tools API ${path} returned ${res.status}`);
  return res.text();
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createTools(userId: number): AgentTool<any>[] {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return [
    {
      name: "get_training_load",
      label: "Get Training Load",
      description:
        "Get current ATL/CTL/TSB training load metrics and a recommendation. Always call this first when the athlete asks about training readiness.",
      parameters: Type.Object({}),
      execute: async () => {
        const text = await apiGet(`/api/tools/training-load?user_id=${userId}`);
        return textResult(text);
      },
    },
    {
      name: "get_recent_activities",
      label: "Get Recent Activities",
      description:
        "Get recent Garmin activities for the last N days. Returns activity type, duration, HR, distance, TSS. Use to understand what the athlete has been doing.",
      parameters: Type.Object({
        days: Type.Optional(
          Type.Number({ description: "Number of days to look back (max 30)", default: 7 })
        ),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { days?: number };
        const days = Math.min(p.days ?? 7, 30);
        const text = await apiGet(
          `/api/tools/recent-activities?user_id=${userId}&days=${days}`
        );
        return textResult(text);
      },
    },
    {
      name: "get_todays_wods",
      label: "Get Today's WODs",
      description:
        "Get today's WODs from all configured CrossFit boxes. Use when generating a training recommendation for today.",
      parameters: Type.Object({}),
      execute: async () => {
        const text = await apiGet(`/api/tools/wods`);
        return textResult(text);
      },
    },
    {
      name: "search_memory",
      label: "Search Semantic Memory",
      description:
        "Search the athlete's semantic memory (PRs, injuries, goals, preferences). Example queries: 'shoulder injury', 'back squat PR', 'goals 2025'",
      parameters: Type.Object({
        query: Type.String({ description: "Search query" }),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { query: string };
        const text = await apiPost(`/api/tools/search-memory`, {
          user_id: userId,
          query: p.query,
        });
        return textResult(text);
      },
    },
    {
      name: "search_episodes",
      label: "Search Episodic Memory",
      description:
        "Search episodic memory for specific events (PRs, achievements, injuries). Use when athlete references past performances.",
      parameters: Type.Object({
        query: Type.String({ description: "Search query" }),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { query: string };
        const text = await apiPost(`/api/tools/search-episodes`, {
          user_id: userId,
          query: p.query,
        });
        return textResult(text);
      },
    },
    {
      name: "get_all_prs",
      label: "Get All PRs",
      description:
        "Get all Personal Records for the athlete. Use when discussing performance history or setting new training goals.",
      parameters: Type.Object({}),
      execute: async () => {
        const text = await apiGet(`/api/tools/prs?user_id=${userId}`);
        return textResult(text);
      },
    },
    {
      name: "save_memory",
      label: "Save Memory",
      description:
        "Save an important fact to semantic memory. Use when learning something important: injuries, PRs, goals, preferences, scaling levels. Categories: 'injury', 'pr', 'goal', 'preference', 'scaling'",
      parameters: Type.Object({
        key: Type.String({ description: "Memory key" }),
        value: Type.String({ description: "Memory value" }),
        category: Type.Optional(
          Type.String({ description: "Category: injury, pr, goal, preference, scaling" })
        ),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { key: string; value: string; category?: string };
        const text = await apiPost(`/api/tools/save-memory`, {
          user_id: userId,
          key: p.key,
          value: p.value,
          category: p.category,
        });
        return textResult(text);
      },
    },
    {
      name: "add_episode",
      label: "Add Episode",
      description:
        "Add an episodic memory entry for a specific event. Use for PRs, injuries, achievements, notable workouts. Categories: 'pr', 'injury', 'achievement', 'scaling', 'wod_result'",
      parameters: Type.Object({
        content: Type.String({ description: "Episode content" }),
        category: Type.Optional(
          Type.String({
            description: "Category: pr, injury, achievement, scaling, wod_result",
          })
        ),
        metadata: Type.Optional(
          Type.Record(Type.String(), Type.Unknown(), {
            description: "Optional metadata e.g. {exercise: 'deadlift', weight_kg: 145}",
          })
        ),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { content: string; category?: string; metadata?: Record<string, unknown> };
        const text = await apiPost(`/api/tools/add-episode`, {
          user_id: userId,
          content: p.content,
          category: p.category,
          metadata: p.metadata,
        });
        return textResult(text);
      },
    },
    {
      name: "get_coaching_profile",
      label: "Get Coaching Profile",
      description:
        "Get the athlete's coaching style profile. Use to adapt communication style (direct/supportive/technical/balanced).",
      parameters: Type.Object({}),
      execute: async () => {
        const text = await apiGet(
          `/api/tools/coaching-profile?user_id=${userId}`
        );
        return textResult(text);
      },
    },
    {
      name: "update_coaching_style",
      label: "Update Coaching Style",
      description:
        "Update the coaching style based on athlete feedback. Valid styles: 'direct', 'supportive', 'technical', 'balanced'",
      parameters: Type.Object({
        coaching_style: Type.Optional(
          Type.String({ description: "Style: direct, supportive, technical, balanced" })
        ),
        notes: Type.Optional(
          Type.String({ description: "Specific communication preferences" })
        ),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { coaching_style?: string; notes?: string };
        const text = await apiPost(`/api/tools/update-coaching-style`, {
          user_id: userId,
          coaching_style: p.coaching_style,
          notes: p.notes,
        });
        return textResult(text);
      },
    },
    {
      name: "get_training_preferences",
      label: "Get Training Preferences",
      description:
        "Get the athlete's free-text training preferences (weekly rhythm, preferred session types, locations). Always call this before planning a week.",
      parameters: Type.Object({}),
      execute: async () => {
        const text = await apiGet(
          `/api/tools/training-preferences?user_id=${userId}`
        );
        return textResult(text);
      },
    },
    {
      name: "save_training_preferences",
      label: "Save Training Preferences",
      description:
        "Overwrite the athlete's training preferences. Use when the athlete describes a lasting change to their weekly training rhythm (not for one-off scheduling constraints).",
      parameters: Type.Object({
        preferences_text: Type.String({
          description: "Complete new preferences text, e.g. '1x/Woche Intervalle auf der 400m Bahn, 2x Laufen, 2x Functional Strength in der Box mit WOD'",
        }),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { preferences_text: string };
        const text = await apiPost(`/api/tools/training-preferences`, {
          user_id: userId,
          preferences_text: p.preferences_text,
        });
        return textResult(text);
      },
    },
    {
      name: "get_week_plan",
      label: "Get Week Plan",
      description:
        "Get the athlete's training plan for a week including every session with its id, date, title, description, status and logged result. Call this before discussing today's training and before logging a result (you need the session id).",
      parameters: Type.Object({
        week_start: Type.Optional(
          Type.String({
            description: "Monday of the week as YYYY-MM-DD. Defaults to the current week.",
          })
        ),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as { week_start?: string };
        const qs = p.week_start ? `&week_start=${p.week_start}` : "";
        const text = await apiGet(`/api/tools/week-plan?user_id=${userId}${qs}`);
        return textResult(text);
      },
    },
    {
      name: "save_week_plan",
      label: "Save Week Plan",
      description:
        "Store the training sessions for a week. Replaces any existing sessions for that week. Every session must be fully written out so the athlete can execute it without asking back.",
      parameters: Type.Object({
        week_start: Type.String({
          description: "Monday of the planned week as YYYY-MM-DD",
        }),
        sessions: Type.Array(
          Type.Object({
            date: Type.String({ description: "Session date as YYYY-MM-DD" }),
            title: Type.String({ description: "Short title, e.g. 'Intervalle 6x400m'" }),
            session_type: Type.Optional(
              Type.String({
                description: "One of: intervals, run, strength, wod, mobility, rest",
              })
            ),
            description: Type.String({
              description:
                "Full session: warm-up, main part with sets/reps/pace/rest, cool-down. Multi-line.",
            }),
            position: Type.Optional(
              Type.Number({ description: "Order within the day, starting at 0" })
            ),
          }),
          { description: "All sessions of the week, in chronological order" }
        ),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as {
          week_start: string;
          sessions: Record<string, unknown>[];
        };
        const text = await apiPost(`/api/tools/week-plan`, {
          user_id: userId,
          week_start: p.week_start,
          sessions: p.sessions,
        });
        return textResult(text);
      },
    },
    {
      name: "log_session_result",
      label: "Log Session Result",
      description:
        "Log the athlete's result for a planned session. Call get_week_plan first to find the matching session id. Use when the athlete reports a completed or skipped workout in chat.",
      parameters: Type.Object({
        session_id: Type.Optional(
          Type.Number({ description: "Session id from get_week_plan (preferred)" })
        ),
        date: Type.Optional(
          Type.String({
            description: "YYYY-MM-DD, only as a fallback when no session id is known",
          })
        ),
        status: Type.Optional(
          Type.String({ description: "'done' or 'skipped' (default: done)" })
        ),
        result_text: Type.Optional(
          Type.String({ description: "Result, e.g. '6x400m, Schnitt 1:31'" })
        ),
        rpe: Type.Optional(
          Type.Number({ description: "Perceived exertion 1-10" })
        ),
        notes: Type.Optional(Type.String({ description: "Additional notes" })),
      }),
      execute: async (_id: string, params: unknown) => {
        const p = params as {
          session_id?: number;
          date?: string;
          status?: string;
          result_text?: string;
          rpe?: number;
          notes?: string;
        };
        const text = await apiPost(`/api/tools/session-result`, {
          user_id: userId,
          session_id: p.session_id,
          date: p.date,
          status: p.status ?? "done",
          result_text: p.result_text,
          rpe: p.rpe,
          notes: p.notes,
        });
        return textResult(text);
      },
    },
  ];
}
